### ISSUE TITLE
multiple anchor preceding section headers willbe removed from final output

### ISSUE THREAD

**Issue by hossman** [asciidoctor/asciidoctor#1874]
According to the user manual (section 30.1: [Defining an Anchor](http://asciidoctor.org/docs/user-manual/#anchordef) ) ...

> In addition to being able to define anchors on sections and blocks, anchors can be defined inline where ever you can type normal text (anchors are a macro substitution). The anchors in the text get replaced with invisible anchor points in the output.

But there seems to be a bug when multiple anchors are defined w/o any normal text between them _and_ they are immediately followed by a section header.

Example:
```
= X Title
:toc:

X X X X X X X X

[[anchor_x2]]

[[anchor_x1]] [[anchor_x0]] Anchor X

X X X X X X X X

[[anchor_a1]]

== [[anchor_a0]] Header A

X X X X X X X X

[[anchor_b2]]

[[anchor_b1]] 
== Header B

X X X X X X X X

[[anchor_c1]]

[[anchor_c2]]

== [[anchor_c0]] Header C

X X X X X X X X

== [[anchor_d0]] Header D

X X X X X X X X
```

html5 backend output (of main content)...
```
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
<div id="anchor_x2" class="paragraph">
<p><a id="anchor_x1"></a> <a id="anchor_x0"></a> Anchor X</p>
</div>
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="anchor_a1"><a id="anchor_a0"></a> Header A</h2>
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="anchor_b1">Header B</h2>
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="anchor_c2"><a id="anchor_c0"></a> Header C</h2>
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="__a_id_anchor_d0_a_header_d"><a id="anchor_d0"></a> Header D</h2>
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
</div>
```

Note that all three `anchor_x#` anchors are included in the HTML output, but for Headers at most one anchor preceeding the `== HEADER` is preserved as the `id` of the resulting `<h#>` tag.

If an anchor is also declared inside of the header, it is also preserved -- and, oddly, as in the case of `anchor_d0` used along with the rest of the header text to define the automatic section id for the `<h#>`, rather then just being used as is for the section id.

In general this behavior with anchors+headers seems inconsistent with the behavior of anchors+text, and contradicts the docs.  All of these anchors should be preserved in the final output.

----

A few final notes:

* I realize this usage of multiple anchors in sequence may seem odd.  FWIW: The situation i'm dealing with is asciidoc files being programatically generated (via pandoc) from the output of another CMS.  Some anchors were created automatically by the CMS, and others were added by users via a special CMS macro in order to get anchors with explicit ids -- but I need to preserve all of them (even when they essentially poin to the same place) since there are existing inbound links to both.
* In the examples above I noted the HTML output created by the html5 backend because it's easy to review the final documents source/structure, but I should note the same problems also occur when using the pdf backend (which is my ultimate goal)
* I've been testing via asciidoctor-ant-1.5.3.jar since that's the latest version i can find on maven.org and ultimately i need to use asciidoctor-ant

**Comment by mojavelinux** [comment]
> for Headers at most one anchor preceding the == HEADER is preserved as the id of the resulting `<h#>` tag.

That is correct. A section can only have one ID, so the last ID wins and becomes the primary anchor.

> If an anchor is also declared inside of the header, it is also preserved -- and, oddly, as in the case of anchor_d0 used along with the rest of the header text to define the automatic section id for the <h#>, rather then just being used as is for the section id.

The ID generation has been fixed in 1.5.7 (see #794).

Only an anchor specified at the end of the section title (offset by a space) is promoted to the ID, and only if an ID is not already specified. In other words, like this:

```
== Header A [[anchor_a0]]
```

> In general this behavior with anchors+headers seems inconsistent with the behavior of anchors+text

That is true. These are different things. Sections are blocks and thus follow the block behavior for IDs / anchors. Inline anchors are more freeform.

> ...and contradicts the docs.

That needs to be fixed. Contributions welcome.

> I realize this usage of multiple anchors in sequence may seem odd.

Actually, I don't think this is odd. There are good reasons why the same element would need multiple reference IDs / aliases. One of those reasons is when the ID changes but you want to preserve a legacy ID. Another is that the information might have multiple labels, such as a canonical version number and a temporal pointer.

It should now be possible to add multiple anchors to the same section title as follows:

```
[#primary-id]
== [[secondary-id]][[tertiary-id]]Section Title
```

or

```
[#primary-id]
== Section Title[[secondary-id]][[tertiary-id]]
```

Note that the anchor element will be placed where the anchor is defined in the AsciiDoc source (it's following the same rules as an inline anchor).

Although we don't sanction it, this could also be specified as follows:

```
== Section Title[[secondary-id]][[tertiary-id]] [[primary-id]]
```

(note the required space before [[primary-id]])

@graphitefriction This can now be added to the docs, but needs to be flagged as 1.5.7 or better.

**Comment by mojavelinux** [comment]
This is now documented in old docs site. That section, as well as the revisions to the Custom ID section, will need to be ported to the new docs site. See https://asciidoctor.org/docs/user-manual/#custom-ids
